curated_transformers/util/auto_model.py

```diff
@@ -0,0 +1,227 @@
+        return decoder
+
+
+class AutoCausalLM(AutoModel[CausalLMModule[KeyValueCache]]):
+    """Causal LM module loaded from the Hugging Face Model Hub."""
+
+    _HF_MODEL_TYPE_TO_CURATED = {
+        "gpt_neox": GPTNeoXCausalLM,
+        "RefinedWeb": RefinedWebModelCausalLM,
+        "RefinedWebModel": RefinedWebModelCausalLM,
+    }
+
+    @classmethod
+    def from_hf_hub(
+        cls,
+        name: str,
+        revision: str = "main",
+        *,
+        device: Optional[torch.device] = None,
+        quantization_config: Optional[BitsAndBytesConfig] = None,
+    ) -> CausalLMModule[KeyValueCache]:
+        causal_lm = _instantiate_module_from_hf_hub(
+            name,
+            revision,
+            device,
+            quantization_config,
+            cls._HF_MODEL_TYPE_TO_CURATED,
+            "causal LM",
+        )
+        assert isinstance(causal_lm, CausalLMModule)
+        return causal_lm
+
+
+class AutoGenerator(AutoModel[GeneratorWrapper]):
+    """Causal LM generator loaded from the Hugging Face Model Hub.
+
+    **NOTE** - This class can currently only be used with following models:
+        - ``databricks/dolly-v2`` variants.
+        - ``tiiuae/falcon`` variants.
+    """
+
+    _DOLLY_V2_PREFIX = "databricks/dolly-v2"
+    _FALCON_PREFIX = "tiiuae/falcon"
+
+    @classmethod
+    def from_hf_hub(
+        cls,
+        name: str,
+        revision: str = "main",
+        *,
+        device: Optional[torch.device] = None,
+        quantization_config: Optional[BitsAndBytesConfig] = None,
+    ) -> GeneratorWrapper:
+        # We need to match the name of the model directly as our
+        # generators bundle model-specific prompts that are specific
+        # to certain fine-tuned models.
+        generator: Optional[GeneratorWrapper] = None
+        if name.startswith(cls._DOLLY_V2_PREFIX):
+            generator = DollyV2Generator.from_hf_hub(
+                name=name,
+                revision=revision,
+                device=device,
+                quantization_config=quantization_config,
+            )
+        elif name.startswith(cls._FALCON_PREFIX):
+            generator = FalconGenerator.from_hf_hub(
+                name=name,
+                revision=revision,
+                device=device,
+                quantization_config=quantization_config,
+            )
+        else:
+            supported_models = (cls._DOLLY_V2_PREFIX, cls._FALCON_PREFIX)
+            raise ValueError(
+                f"Unsupported generator model `{name}`. "
+                f"Supported model: `{supported_models}` variants"
+            )
+
+        assert generator is not None
+        assert isinstance(generator, GeneratorWrapper)
+        return generator
+
+
+def _instantiate_module_from_hf_hub(
+    name: str,
+    revision: str,
+    device: Optional[torch.device],
+    quantization_config: Optional[BitsAndBytesConfig],
+    model_type_to_class_map: Mapping[str, Type],
+    module_type: str,
```

Comment:
I am not a big fan of this stringly-typed argument. Maybe we could make this a class method of `AutoModel`, so that we could say `AutoDecoder`/`AutoEncoder`/`AutoCausalLM` below?

Reply:
Yeah, I agree that it's iffy. Will add a class method.